### PR TITLE
Support "static" __getitem__ on Numba types in `@njit` code.

### LINKED ISSUE #5748: Support getitem for type types.
The syntax for an array uses square brackets, for example: `int64[:]`. Numba currently does not support this syntax within an `njit`context. For example the following will not work:

```python
from numba import njit
from numba.types import int64
from numba.typed import List
import numpy as np


@njit
def foo():
    l = List.empty_list(int64[:])
    l.append(np.zeros(23))
    return l


foo()
```

And will instead error out with:

```
Traceback (most recent call last):
  File "/Users/vhaenel/git/numba/numba/core/typeinfer.py", line 155, in propagate
    constraint(typeinfer)
  File "/Users/vhaenel/git/numba/numba/core/typeinfer.py", line 398, in __call__
    self.fallback(typeinfer)
  File "/Users/vhaenel/git/numba/numba/core/typeinfer.py", line 618, in __call__
    self.resolve(typeinfer, typeinfer.typevars, fnty=fnty)
  File "/Users/vhaenel/git/numba/numba/core/typeinfer.py", line 515, in resolve
    sig = typeinfer.resolve_call(fnty, pos_args, kw_args)
  File "/Users/vhaenel/git/numba/numba/core/typeinfer.py", line 1459, in resolve_call
    return self.context.resolve_function_type(fnty, pos_args, kw_args)
  File "/Users/vhaenel/git/numba/numba/core/typing/context.py", line 214, in resolve_function_type
    raise last_exception
  File "/Users/vhaenel/git/numba/numba/core/typing/context.py", line 197, in resolve_function_type
    res = self._resolve_user_function_type(func, args, kws)
  File "/Users/vhaenel/git/numba/numba/core/typing/context.py", line 249, in _resolve_user_function_type
    return func.get_call_type(self, args, kws)
  File "/Users/vhaenel/git/numba/numba/core/types/functions.py", line 148, in get_call_type
    failures.raise_error()
  File "/Users/vhaenel/git/numba/numba/core/types/functions.py", line 77, in raise_error
    raise errors.TypingError(self.format())
numba.core.errors.TypingError: Invalid use of Function(<built-in function getitem>) with argument(s) of type(s): (class(int64), slice<a:b>)
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
In definition 2:
    All templates rejected with literals.
In definition 3:
    All templates rejected without literals.
In definition 4:
    All templates rejected with literals.
In definition 5:
    All templates rejected without literals.
In definition 6:
    All templates rejected with literals.
In definition 7:
    All templates rejected without literals.
In definition 8:
    All templates rejected with literals.
In definition 9:
    All templates rejected without literals.
In definition 10:
    All templates rejected with literals.
In definition 11:
    All templates rejected without literals.
In definition 12:
    All templates rejected with literals.
In definition 13:
    All templates rejected without literals.
In definition 14:
    All templates rejected with literals.
In definition 15:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: typing of intrinsic-call at foo070.py (9)
[2] During: typing of static-get-item at foo070.py (9)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "foo070.py", line 14, in <module>
    foo()
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 401, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 342, in error_rewrite
    raise e
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 353, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "/Users/vhaenel/git/numba/numba/core/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 794, in compile
    cres = self._compiler.compile(args, return_type)
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 81, in compile
    raise retval
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 91, in _compile_cached
    retval = self._compile_core(args, return_type)
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 109, in _compile_core
    pipeline_class=self.pipeline_class)
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 568, in compile_extra
    return pipeline.compile_extra(func)
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 339, in compile_extra
    return self._compile_bytecode()
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 401, in _compile_bytecode
    return self._compile_core()
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 381, in _compile_core
    raise e
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 372, in _compile_core
    pm.run(self.state)
  File "/Users/vhaenel/git/numba/numba/core/compiler_machinery.py", line 341, in run
    raise patched_exception
  File "/Users/vhaenel/git/numba/numba/core/compiler_machinery.py", line 332, in run
    self._runPass(idx, pass_inst, state)
  File "/Users/vhaenel/git/numba/numba/core/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/Users/vhaenel/git/numba/numba/core/compiler_machinery.py", line 291, in _runPass
    mutated |= check(pss.run_pass, internal_state)
  File "/Users/vhaenel/git/numba/numba/core/compiler_machinery.py", line 264, in check
    mangled = func(compiler_state)
  File "/Users/vhaenel/git/numba/numba/core/typed_passes.py", line 98, in run_pass
    raise_errors=self._raise_errors)
  File "/Users/vhaenel/git/numba/numba/core/typed_passes.py", line 70, in type_inference_stage
    infer.propagate(raise_errors=raise_errors)
  File "/Users/vhaenel/git/numba/numba/core/typeinfer.py", line 994, in propagate
    raise errors[0]
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<built-in function getitem>) with argument(s) of type(s): (class(int64), slice<a:b>)
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
In definition 2:
    All templates rejected with literals.
In definition 3:
    All templates rejected without literals.
In definition 4:
    All templates rejected with literals.
In definition 5:
    All templates rejected without literals.
In definition 6:
    All templates rejected with literals.
In definition 7:
    All templates rejected without literals.
In definition 8:
    All templates rejected with literals.
In definition 9:
    All templates rejected without literals.
In definition 10:
    All templates rejected with literals.
In definition 11:
    All templates rejected without literals.
In definition 12:
    All templates rejected with literals.
In definition 13:
    All templates rejected without literals.
In definition 14:
    All templates rejected with literals.
In definition 15:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: typing of intrinsic-call at foo070.py (9)
[2] During: typing of static-get-item at foo070.py (9)

File "foo070.py", line 9:
def foo():
    l = List.empty_list(int64[:])
    ^
```
 

Implementing `getitem` for type types should resolve this.

### LINKED ISSUE #7683: connot create array types inside jitted functions
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

as title

```python
import numba
from numba.typed.typedlist import List


@numba.njit
def test_typing_1():
    ls = List.empty_list(numba.float64[:])


f8_1d_type = numba.float64[:]


@numba.njit
def test_typing_2():
    ls = List.empty_list(f8_1d_type)


test_typing_1()
test_typing_2()
```
```
No implementation of function Function(<built-in function getitem>) found for signature:
 
 >>> getitem(class(float64), slice<a:b>)
 
There are 22 candidate implementations:
    - Of which 22 did not match due to:
    Overload of function 'getitem': File: <numerous>: Line N/A.
      With argument(s): '(class(float64), slice<a:b>)':
     No match.
During: typing of intrinsic-call at <ipython-input-3-88edda07e39d> (3)
During: typing of static-get-item at <ipython-input-3-88edda07e39d> (3)
File "<ipython-input-3-88edda07e39d>", line 3:
def test_typing_1():
    ls = List.empty_list(numba.float64[:])
```
a MWE would be:
```python
@numba.njit
def test_typing():
    numba.float64[:]
```

### LINKED ISSUE #8808: jitclass doesn't work when dict class attribute has array as value type
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Trying to `jitclass` a class that has a `Dict` attribute with an array as the value type fails. Code and traceback below. Removing `jitclass` results in code that runs fine.

```python
import numpy as np
from numba import int64
from numba.experimental import jitclass
from numba.typed import Dict
from numba.types import DictType


@jitclass(
    [
        ("dict", DictType(int64, int64[:, :])),
    ]
)
class StructureManager:
    def __init__(self):
        self.dict = Dict.empty(key_type=int64, value_type=int64[:, :])


s = StructureManager()
s.dict[0] = np.zeros((2, 2), dtype=int)
```

```python
Traceback (most recent call last):
  File "/Users/dstansby/projects/brainglobe/cellfinder-core/test.py", line 18, in <module>
    s = StructureManager()
  File "/Users/dstansby/mambaforge/envs/cellfinder-core/lib/python3.10/site-packages/numba/experimental/jitclass/base.py", line 124, in __call__
    return cls._ctor(*bind.args[1:], **bind.kwargs)
  File "/Users/dstansby/mambaforge/envs/cellfinder-core/lib/python3.10/site-packages/numba/core/dispatcher.py", line 468, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/dstansby/mambaforge/envs/cellfinder-core/lib/python3.10/site-packages/numba/core/dispatcher.py", line 409, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<built-in function getitem>) found for signature:
 
 >>> getitem(class(int64), UniTuple(slice<a:b> x 2))
 
There are 22 candidate implementations:
  - Of which 22 did not match due to:
  Overload of function 'getitem': File: <numerous>: Line N/A.
    With argument(s): '(class(int64), UniTuple(slice<a:b> x 2))':
   No match.

During: typing of intrinsic-call at /Users/dstansby/projects/brainglobe/cellfinder-core/test.py (15)
During: typing of static-get-item at /Users/dstansby/projects/brainglobe/cellfinder-core/test.py (15)

File "test.py", line 15:
    def __init__(self):
        self.dict = Dict.empty(key_type=int64, value_type=int64[:, :])
        ^

During: resolving callee type: jitclass.StructureManager#115e0b7f0<dict:DictType[int64,array(int64, 2d, A)]<iv=None>>
During: typing of call at <string> (3)

During: resolving callee type: jitclass.StructureManager#115e0b7f0<dict:DictType[int64,array(int64, 2d, A)]<iv=None>>
During: typing of call at <string> (3)


File "<string>", line 3:
<source missing, REPL/exec in use?>
```

### PR DESCRIPTION
**Pull request by stuartarchibald**
This adds support for the `__getitem__` operation on Numba types within `@njit` compiled regions. i.e. it makes this sort of thing possible:

```python
from numba import njit, types, typed

@njit
def foo():
    return typed.List.empty_list(types.float64[:, ::1])

foo()
```

The logic in the Type metaclass is relatively relaxed over what's acceptable in a `__getitem__`, this patch is necessarily more restrictive and the compilable `__getitem__` is implemented only for types that are practical for this sort of use i.e. those in the numerical type system.

Tests are added and docs are updated to note that this is now supported.

Closes #8808
Closes #7683
Closes #5748

### CODE REVIEW

File: docs/source/reference/types.rst [hunk 1]
```diff
@@ -102,6 +102,17 @@ the beginning or the end of the index specification::
    >>> numba.float32[::1, :, :]
    array(float32, 3d, F)
 
+This style of type declaration is supported within Numba :term:`JIT` compiled
```
Comment (gmarkall) [inline docs/source/reference/types.rst:105]: This links to docs that make a distinction between JIT compilation and AOT compilation, which I think is misleading in this context. Would it be better if it just didn't mention "JIT" and said "Numba-compiled functions" instead?

File: docs/source/reference/types.rst [hunk 2]
```diff
@@ -102,6 +102,17 @@ the beginning or the end of the index specification::
...
+functions e.g. declaring the type of a :ref:`typed.List <feature-typed-list>`.
```
Comment (gmarkall) [inline docs/source/reference/types.rst:106]: I think this could do with a comma to separate the clauses of the sentence

```suggestion
functions, e.g. declaring the type of a :ref:`typed.List <feature-typed-list>`.
```

File: numba/core/typing/builtins.py [hunk 3]
```diff
@@ -654,6 +654,26 @@ def generic(self, args, kws):
             sig = signature(ret, *args)
             return sig
 
+@infer
+class StaticGetItemClass(AbstractTemplate):
+    """This handles the "static_getitem" that occurs when a Numba type has a
+    __getitem__ made on it in input source e.g:
```
Comment (gmarkall) [inline numba/core/typing/builtins.py:660]: I found this a bit hard to parse and understand - is this simplification clearer and easier?

```suggestion
    """This handles the "static_getitem" when a Numba type is subscripted,
    e.g.:
```

File: docs/source/reference/types.rst [hunk 4]
```diff
@@ -102,6 +102,17 @@ the beginning or the end of the index specification::
...
+::
+
+    from numba import njit, types, typed
+
+    @njit
+    def example():
+        return typed.List.empty_list(types.float64[:, ::1])
+
+
```
Comment (gmarkall) [inline docs/source/reference/types.rst:115]: Perhaps add a note here that only simple numerical types, and not compound types, are supported?

File: numba/core/typing/builtins.py [hunk 5]
```diff
@@ -654,6 +654,26 @@ def generic(self, args, kws):
...
+    var = typed.List.empty_list(float64[::1, :])
+    it only allows this on simple numerical types i.e. compound types, like
```
Comment (gmarkall) [inline numba/core/typing/builtins.py:662]: Simplifying again:

```suggestion
    It only allows this on simple numerical types. Compound types, like
```

File: numba/np/arrayobj.py [hunk 6]
```diff
@@ -2922,6 +2922,31 @@ def dtype_type(context, builder, dtypety, dtypeval):
...
 
 
+# ------------------------------------------------------------------------------
+# static_getitem on Numba numerical types to create "array" types
+
+
+@lower_builtin('static_getitem', types.Boolean, types.Any)
+@lower_builtin('static_getitem', types.NumberClass, types.Any)
+def static_getitem_number_clazz(context, builder, sig, args):
+    """This handles the "static_getitem" that occurs when a Numba type a
```
Comment (gmarkall) [inline numba/np/arrayobj.py:2932]: Similar comments on this docstring to those above in the typing.

File: numba/core/typing/builtins.py [hunk 7]
```diff
@@ -654,6 +654,26 @@ def generic(self, args, kws):
...
+    records, are not supported.
+    """
+    key = "static_getitem"
+
+    def generic(self, args, kws):
+        clazz, idx = args
+        accept = (types.NumberClass, types.Boolean)
+        if not isinstance(clazz, accept):
+            return
+        ret = clazz.dtype[idx]
```
Comment (gmarkall) [inline numba/core/typing/builtins.py:672]: TBC: I'm not yet sure how this would work if `clazz` is a `Boolean`, as that doesn't appear to have a `dtype`.

File: numba/tests/test_getitem_on_types.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
+import unittest
+from itertools import product
+from numba import types, njit, typed, errors
+from numba.tests.support import TestCase
+
+
+class TestGetitemOnTypes(TestCase):
+    # Tests getitem on the type types.
+
+    def test_static_getitem_on_type(self):
```
Comment (gmarkall) [inline numba/tests/test_getitem_on_types.py:10]: TBC: This test looks OK, I still need to look at the remaining tests.

**Review by gmarkall — COMMENTED** [review]
I've only had chance to partially review this so I am leaving a comments review with what I had accumulated on the diff so far. In general I think the idea and implementation are fine and it's just about me understanding all the details.

Comments that denote something I need to go and look at when I finish the review are marked "TBC".

File: numba/core/typing/builtins.py (hunk 7 above)
Comment (stuartarchibald) [reply numba/core/typing/builtins.py:672]: I think your assertion is correct, use of `types.Boolean` is incorrect/superfluous, as `clazz` would be an instance of a `types.functions.NumberClass` and "boolean" types would work through that.

Comment (gmarkall) [reply numba/core/typing/builtins.py:672]: OK - does that just need removing then?

File: numba/tests/test_getitem_on_types.py (hunk 8 above)
Comment (gmarkall) [reply numba/tests/test_getitem_on_types.py:10]: Update: the remaining tests look good to me.

File: numba/core/typing/builtins.py (hunk 7 above)
Comment (stuartarchibald) [reply numba/core/typing/builtins.py:670]: Yes, I think so, 3c5da76 contains this change.

**Comment by stuartarchibald** [comment]
@gmarkall many thanks for the review. RE: "other minor stuff", I agree with and have implemented the suggestions in 3c5da76.

**Review by gmarkall — APPROVED** [review]
Thanks for the updates, this looks good!
